diff --git a/.gitignore b/.gitignore
index f73f35f..4983756 100644
--- a/.gitignore
+++ b/.gitignore
@@ -100,6 +100,8 @@ x509.genkey
 #
 configs/*.config
 localversion
+redhat/kabi/kabi-datafiles.list
+*.kb_dat
 
 # configuration file for rh- targets
 .rhel7git.mk
diff --git a/Makefile b/Makefile
index fc7a8cd..3ba0340 100644
--- a/Makefile
+++ b/Makefile
@@ -345,6 +345,10 @@ INSTALLKERNEL  := installkernel
 DEPMOD		= /sbin/depmod
 PERL		= perl
 CHECK		= sparse
+KABIPARSER	= /usr/sbin/kabi-parser
+
+KABIDATADIR	= redhat/kabi
+KABIDATAFILE	= $(KABIDATADIR)/kabi-datafiles.list
 
 CHECKFLAGS     := -D__linux__ -Dlinux -D__STDC__ -Dunix -D__unix__ \
 		  -Wbitwise -Wno-return-void $(CF)
@@ -408,7 +412,7 @@ KERNELVERSION = $(VERSION)$(if $(PATCHLEVEL),.$(PATCHLEVEL)$(if $(SUBLEVEL),.$(S
 export VERSION PATCHLEVEL SUBLEVEL KERNELRELEASE KERNELVERSION
 export ARCH SRCARCH CONFIG_SHELL HOSTCC HOSTCFLAGS CROSS_COMPILE AS LD CC
 export CPP AR NM STRIP OBJCOPY OBJDUMP
-export MAKE AWK GENKSYMS INSTALLKERNEL PERL UTS_MACHINE
+export MAKE AWK GENKSYMS INSTALLKERNEL PERL UTS_MACHINE KABIPARSER
 export HOSTCXX HOSTCXXFLAGS LDFLAGS_MODULE CHECK CHECKFLAGS
 
 export KBUILD_CPPFLAGS NOSTDINC_FLAGS LINUXINCLUDE OBJCOPYFLAGS LDFLAGS
diff --git a/redhat/Makefile b/redhat/Makefile
index 07783d8..a7dac25 100644
--- a/redhat/Makefile
+++ b/redhat/Makefile
@@ -57,6 +57,10 @@ rh-configs-prep: rh-configs
 rh-configs-arch-prep:
 	cd $(REDHAT)/configs; ./build_configs.sh $(MACH)
 
+rh-clean-kabi:
+	rm -f $(REDHAT)/kabi/kabi-datafiles.list
+	find $(TOPDIR) \( -iname '*.kb_dat' -or -iname '*.i' \) -type f | xargs rm -f
+
 rh-clean-configs:
 	cd $(REDHAT)/configs; rm -f kernel-*.config
 
@@ -71,7 +75,7 @@ rh-clean-rpmdirs:
 		rm -rf $$i; \
 	done;
 
-rh-clean: rh-clean-sources rh-clean-configs rh-clean-rpmdirs
+rh-clean: rh-clean-sources rh-clean-configs rh-clean-rpmdirs rh-clean-kabi
 
 rh-stub-key:
 	@echo "Copying pre-generated keys";
diff --git a/scripts/Makefile.build b/scripts/Makefile.build
index d5d859c..1a4db5e 100644
--- a/scripts/Makefile.build
+++ b/scripts/Makefile.build
@@ -209,6 +209,18 @@ cmd_cc_i_c       = $(CPP) $(c_flags)   -o $@ $<
 $(obj)/%.i: $(src)/%.c FORCE
 	$(call if_changed_dep,cc_i_c)
 
+cmd_kabiparser = 					\
+    $(CPP) $(c_flags) -o $(1) $<;			\
+    echo $(2) >> redhat/kabi/kabi-datafiles.list;	\
+    $(KABIPARSER) -xf $(2) $(1);
+
+quiet_cmd_cc_kabiparser_c = KABI $@
+cmd_cc_kabiparser_c = 					\
+    $(call cmd_kabiparser, $(@:.kb_dat=.i), $@)
+
+$(obj)/%.kb_dat : $(src)/%.c FORCE
+	$(call cmd,cc_kabiparser_c)
+
 cmd_gensymtypes =                                                           \
     $(CPP) -D__GENKSYMS__ $(c_flags) $< |                                   \
     $(GENKSYMS) $(if $(1), -T $(2))                                         \
@@ -253,6 +265,8 @@ cmd_modversions =								\
 		$(call cmd_gensymtypes,$(KBUILD_SYMTYPES),$(@:.o=.symtypes))	\
 		    > $(@D)/.tmp_$(@F:.o=.ver);					\
 										\
+		$(call cmd_kabiparser,$(@:.o=.i),$(@:.o=.kb_dat))		\
+										\
 		$(LD) $(LDFLAGS) -r -o $@ $(@D)/.tmp_$(@F) 			\
 			-T $(@D)/.tmp_$(@F:.o=.ver);				\
 		rm -f $(@D)/.tmp_$(@F) $(@D)/.tmp_$(@F:.o=.ver);		\
